sled-agent/src/metrics.rs

```diff
@@ -4,124 +4,415 @@
 
 //! Metrics produced by the sled-agent for collection by oximeter.
 
+use illumos_utils::running_zone::RunningZone;
 use omicron_common::api::internal::nexus::ProducerEndpoint;
 use omicron_common::api::internal::nexus::ProducerKind;
-use oximeter::types::MetricsError;
-use oximeter::types::ProducerRegistry;
-use oximeter_instruments::kstat::link;
+use omicron_common::api::internal::shared::SledIdentifiers;
+use oximeter_instruments::kstat::link::sled_data_link::SledDataLink;
 use oximeter_instruments::kstat::CollectionDetails;
 use oximeter_instruments::kstat::Error as KstatError;
 use oximeter_instruments::kstat::KstatSampler;
 use oximeter_instruments::kstat::TargetId;
 use oximeter_producer::LogConfig;
 use oximeter_producer::Server as ProducerServer;
-use sled_hardware_types::Baseboard;
 use slog::Logger;
+use std::collections::btree_map::Entry;
 use std::collections::BTreeMap;
 use std::net::Ipv6Addr;
 use std::net::SocketAddr;
-use std::sync::Arc;
-use std::sync::Mutex;
 use std::time::Duration;
+use tokio::sync::mpsc;
 use uuid::Uuid;
 
 /// The interval on which we ask `oximeter` to poll us for metric data.
-pub(crate) const METRIC_COLLECTION_INTERVAL: Duration = Duration::from_secs(30);
+const METRIC_COLLECTION_INTERVAL: Duration = Duration::from_secs(30);
 
 /// The interval on which we sample link metrics.
-pub(crate) const LINK_SAMPLE_INTERVAL: Duration = Duration::from_secs(10);
+//
+// TODO(https://github.com/oxidecomputer/omicron/issues/5695)
+// These should probably be sampled much densely. We may want to wait for
+// https://github.com/oxidecomputer/omicron/issues/740, which would handle
+// pagination between the producer and collector, as sampling at < 1s for many
+// links could lead to quite large requests. Or we can eat the memory cost for
+// now.
+const LINK_SAMPLE_INTERVAL: Duration = Duration::from_secs(10);
```

Comment:
TODO aside, should we make this configurable somewhere? Or is the worry that it can be set too low (though we could guard against that)?

Reply:
I'm not sure. I kind of like keeping it private here, where we can include notes like this about implementation details.